src/gradient/gradient.class.ts

```diff
@@ -1,493 +1,334 @@
-        y2: options.coords.y2 || 0
-      };
-
-      if (this.type === 'radial') {
-        coords.r1 = options.coords.r1 || 0;
-        coords.r2 = options.coords.r2 || 0;
-      }
-
-      this.coords = coords;
-      this.colorStops = options.colorStops.slice();
-    },
+  toSVG(object: FabricObject, { additionalTransform: preTransform }: { additionalTransform?: string } = {}) {
+    const markup = [],
+      transform = this.gradientTransform ? this.gradientTransform.concat() : iMatrix.concat(),
+      gradientUnits = this.gradientUnits === 'pixels' ? 'userSpaceOnUse' : 'objectBoundingBox';
+    // colorStops must be sorted ascending
+    const colorStops = this.colorStops.concat().sort((a, b) => {
+      return a.offset - b.offset;
+    });
 
-    /**
-     * Adds another colorStop
-     * @param {Object} colorStop Object with offset and color
-     * @return {fabric.Gradient} thisArg
-     */
-    addColorStop: function(colorStops) {
-      for (var position in colorStops) {
-        var color = new Color(colorStops[position]);
-        this.colorStops.push({
-          offset: parseFloat(position),
-          color: color.toRgb(),
-          opacity: color.getAlpha()
+    let offsetX = -this.offsetX, offsetY = -this.offsetY;
+    if (gradientUnits === 'objectBoundingBox') {
+      offsetX /= object.width;
+      offsetY /= object.height;
+    }
+    else {
+      offsetX += object.width / 2;
+      offsetY += object.height / 2;
+    }
+    if (object.type === 'path' && this.gradientUnits !== 'percentage') {
+      offsetX -= object.pathOffset.x;
+      offsetY -= object.pathOffset.y;
+    }
+    transform[4] -= offsetX;
+    transform[5] -= offsetY;
+
+    const commonAttributes = [
+      `id="SVGID_${this.id}"`,
+      `gradientUnits="${gradientUnits}"`,
+      `gradientTransform="${preTransform ? preTransform + ' ' : ''}${matrixToSVG(transform)}"`,
+      ''
+    ].join(' ');
+
+    if (this.type === 'linear') {
+      const { x1, y1, x2, y2 } = this.coords;
+      markup.push(
+        '<linearGradient ',
+        commonAttributes,
+        ' x1="', x1,
+        '" y1="', y1,
+        '" x2="', x2,
+        '" y2="', y2,
+        '">\n'
+      );
+    }
+    else if (this.type === 'radial') {
+      const { x1, y1, x2, y2, r1, r2 } = this.coords as GradientCoords<'radial'>;
+      const needsSwap = r1 > r2;
+      // svg radial gradient has just 1 radius. the biggest.
+      markup.push(
+        '<radialGradient ',
+        commonAttributes,
+        ' cx="', needsSwap ? x1 : x2,
+        '" cy="', needsSwap ? y1 : y2,
+        '" r="', needsSwap ? r1 : r2,
+        '" fx="', needsSwap ? x2 : x1,
+        '" fy="', needsSwap ? y2 : y1,
+        '">\n'
+      );
+      if (needsSwap) {
+        // svg goes from internal to external radius. if radius are inverted, swap color stops.
+        colorStops.reverse(); //  mutates array
+        colorStops.forEach(colorStop => {
+          colorStop.offset = 1 - colorStop.offset;
```

Comment:
i would argue this is a mutation too.
As it was in the old code and i wonder how it does not break the code gradient after an svg export.
i will try to write a tests that make this fail, and then fix it.

i think the safest one would be:
`colorStops  = colorStops.map((colorStop) => ({ ...colorStop, offset: 1 -colorStop.offset }))`

Reply:
> As it was in the old code and i wonder how it does not break the code gradient after an svg export.

I agree.

> i would argue this is a mutation too.

Disagree since `colorStops` is a local copy

Reply:
is still mutation because colorStops is a local copy of the array, not of the object contained inside. Those are still references to the original objects.

Reply:
Ohh shit

Reply:
Why didn't you commit?